### ISSUE TITLE
OAuth - Error handling auth response. Error: invalid_grant

### ISSUE THREAD

**Issue by kimfucious** [aws-amplify/amplify-js#6041]
This is a weird one, but I've exhausted myself trying to get to the bottom of it, so I'm asking it here, thinking it is a bug, due to the very specific behaviour.

The behaviour is similar to [#3185](https://github.com/aws-amplify/amplify-js/issues/3185), but different, I think.

- I've got a React web app that uses Amplify Auth for user authentication
- users can sign up and sign in with a username and password
- there is a preSignup script that will link the social users to the username/password user (if it exists).
- at present, initial user creation works.
- if a user tries to sign in with Google or Apple before the initial username/password user is created, the preSignup script will return a response, that the app acts on to inform the user to create a username/password user first.

After the initial username/password user has been created, a user *can* sign in using `Auth.federatedSignIn({ provider: "SignInWithApple"});`.  It works smashingly! (even though I know it's not supported officially by Amplify yet).

However, after the initial username/password user has been created, if the user tries to log in with Google, before attempting to sign in with Apple, via `Auth.federatedSignIn({ provider: "Google"});`, an error is thrown when trying to do a post to ` https://myapp.auth.[region].amazoncognito.com/oauth2/token`.

The error message is:
```console
OAuth - Error handling auth response. Error: invalid_grant
```
What appears to happen is: the user is returned to the app via the user pool callbackUrl, at which point Hub receives this and says:
```js
"Retrieving tokens from https://myapp.auth.[region].amazoncognito.com/oauth2/token"
```
That's when the error is thrown.  And since the error is thrown, the preSignup script is never called (so says CloudWatch), and the app is stuck in the water.

I understand this error means that one of the following has occurred:
- Refresh token has been revoked.
- Authorization code has been consumed already or does not exist.

For the record, I am not calling `Amplify.configure(awsconfig);` twice anywhere.  It's called once in `index.js`.

**_Now here's the weird part..._**

I can only sign in with Google after first signing in with Apple (after the initial username/password user is created, as mentioned prior).

In other words, after I create the initial username/password user, I can immediately (without ever signing in with username/password) sign in with Apple, but I can't with Google.

But after I sign in with Apple, I then can sign in with Google.

And I got no idea why 🤷 

I realize this is weird and that you may need more info, so please ask for what you may need, and I'll be happy to share.

**Comment by kimfucious** [comment]
I've made an interesting discovery, which I'll share, but first a little background.

This user pool serves both a web-app and a react-native app (Expo).  There are actually three user pools for : dev, staging, and prod.

This means that the callback and sign-out urls are mixed, like the below:

**dev:**
callbacks:  exp://192.168.1.116:19000/--/, https://localhost:3000/auth-callback
sign-outs: exp://192.168.1.116:19000/--/, https://localhost:3000

**staging:**
callbacks:  exp://exp.host/@kimfucious/myapp-staging, https://staging.myapp.com/auth-callback
sign-outs: exp://exp.host/@kimfucious/myapp-staging, https://staging.myapp.com

**production:**
callbacks:  myapp://, https://myapp.com/auth-callback
sign-outs: myapp://, https://myapp.com

My reporting of this issue is based on the web-app's behavior in dev, at the moment.

I do use a manual configuration of Amplify, so that the web app and react-native app only use the callback and sign-out urls that are pertinent to them, so in the web-app, this would look like this as part of a larger config:
```js
...
oauth: {
domain: "myapp-dev.auth.[region].amazoncognito.com",
scope: [
    "phone",
    "email",
    "openid",
    "profile",
    "aws.cognito.signin.user.admin"
],
redirectSignIn: "https://localhost:3000/auth-callback",
redirectSignOut: "https://localhost:3000",
responseType: "code"
},
...
```
My discovery is that by changing the order of the callback urls in the dev user pool, like the below, will fix the strange behavior that I mentioned prior:

**dev:**
callbacks:  https://localhost:3000/auth-callback, exp://192.168.1.116:19000/--/
sign-outs: https://localhost:3000, exp://192.168.1.116:19000/--/ 

The behavior is still strange, because it's very specific.  And only happens in a certain scenarios that I described prior.

I've yet to try this on the RN app, but since this is "kinda" fixed, I'll be getting to that soon.

I still believe that this may be a bug, so I'm leaving this open for now.

**Comment by tvergho** [comment]
I'm also getting a similar issue – creating an initial user through Facebook works fine, but using the preSignUp script to link an existing user with a social account results in the `OAuth - Error handling auth response. Error: invalid_grant` error. Would like to hear about any solutions to this!

**Comment by Jun711** [comment]
@kimfucious 
I got this same error through different steps. Refer to the above linked issue.

1. login with google (this gmail doesn't exist in user pool yet)
2. delete google account 
3. sign up cognito using the same gmail
4. login with google (link account)

**Comment by rahulje9** [comment]
I had this issue, this happened when I tried to log in after signing out from the app, how I got this fixed was while signing out I called `Auth.signOut()` which solved the issue for me.

**Comment by rahulje9** [comment]
can you please share the `preSignup script` as I want to merge the social login with the email and password login.

**Comment by kimfucious** [comment]
Hi @rahulje9 

My pre-signup script is somewhat of a mess and probably would confuse you more than help.

Checkout the answer [here](https://stackoverflow.com/questions/59635482/aws-cognito-best-practice-to-handle-same-user-with-same-email-address-signing).  It might get you on the right path with what you're trying to do.

**Comment by sammartinez** [comment]
Hey @kimfucious, as you have found and what we recommend to solve this use case is around the pre signup scripts. We do, in the Amplify CLI, have [Lambda Triggers](https://docs.amplify.aws/cli/usage/lambda-triggers) that you are able to configure against `pre-signup` in order to use. Since this is what we would recommend to use, I am going to resolve this issue. Please let us know if there is anymore questions you have in regards to this. Thanks!
